src/MongoDB.Driver.Core/Core/Misc/ExceptionMapper.cs

```diff
@@ -118,17 +118,37 @@ public static Exception Map(ConnectionId connectionId, WriteConcernResult writeC
         /// <returns>The exception, or null if no exception necessary.</returns>
         public static Exception MapNotPrimaryOrNodeIsRecovering(ConnectionId connectionId, BsonDocument response, string errorMessageFieldName)
         {
+            BsonValue codeBsonValue;
+            if (response.TryGetValue("code", out codeBsonValue) && codeBsonValue.IsNumeric)
+            {
+                var code = (ServerErrorCode)codeBsonValue.ToInt32();
+                switch (code)
+                {
+                    case ServerErrorCode.NotMaster:
+                    case ServerErrorCode.NotMasterNoSlaveOk:
+                        return new MongoNotPrimaryException(connectionId, response);
+
+                    case ServerErrorCode.InterruptedAtShutdown:
+                    case ServerErrorCode.InterruptedDueToReplStateChange:
+                    case ServerErrorCode.NotMasterOrSecondary:
+                    case ServerErrorCode.PrimarySteppedDown:
+                    case ServerErrorCode.ShutdownInProgress:
+                        return new MongoNodeIsRecoveringException(connectionId, response);
+                }
+            }
+
             BsonValue errorMessageBsonValue;
             if (response.TryGetValue(errorMessageFieldName, out errorMessageBsonValue) && errorMessageBsonValue.IsString)
             {
                 var errorMessage = errorMessageBsonValue.ToString();
-                if (errorMessage.StartsWith("not master", StringComparison.OrdinalIgnoreCase))
+                if (errorMessage.IndexOf("node is recovering", StringComparison.OrdinalIgnoreCase) != -1 ||
```

Comment:
There is no case insensitive version of Contains so IndexOf is the closest equivalent.